### PR TITLE
Feature: Multi-file support for Credential Types

### PR DESCRIPTION
**Pull request by jladdjr**
Connect #349

### CODE REVIEW

File: awx/main/fields.py [hunk 1]
```diff
@@ -740,8 +739,24 @@ def validate(self, value, model_instance):
...
+            raise django_exceptions.ValidationError(
+                _('Must use multi-file syntax when injecting multiple files'),
+                code='invalid',
+                params={'value': value},
+            )
+        if 'template' not in template_names:
+            valid_namespace['tower'].filename = TowerNamespace()
+            for template_name in template_names:
+                template_name = template_name.split('.')[1]
+                setattr(valid_namespace['tower'].filename, template_name, 'EXAMPLE')
```
Comment (AlanCoding) [inline awx/main/fields.py:765]: Does this validate if the user uses syntax like `"{{tower.filename.cert}"`, where the cert file isn't specified, and is that even possible to check?

Comment (jladdjr) [reply awx/main/fields.py:765]: @AlanCoding - Good question - the short answer is that lines 760-771 (in the new version of this file) will throw a `ValidationError` if a file template is not defined. 

The slightly longer answer is.. in the section I mentioned, lines 763-765 simulate rendering a template using information in a namespace. If `tower.filename.cert` is not in the namespace, then `ValidationError` will be raised when `render` is called. Lines 754-758 create the namespace (using `EXAMPLE` in place of the actual file template). So, in short, we simulate rendering extravars, environment variables, and file templates in order to confirm that all the variables they require exist.

File: awx/main/fields.py [hunk 2]
```diff
@@ -686,11 +686,10 @@ def schema(self, model_instance):
             'properties': {
                 'file': {
                     'type': 'object',
-                    'properties': {
-                        'template': {'type': 'string'},
+                    'patternProperties': {
+                        '^template(\.[a-zA-Z_]+)?$': {'type': 'string'},
```
Comment (ryanpetrello) [inline awx/main/fields.py]: Technically speaking, this _could be_ any valid jinja variable name.  Instead of:

```regex
^template(\.[a-zA-Z_]+)?$
```

...you probably want to match the pattern we use for the input definitions: https://github.com/ansible/awx/blob/96904968d850c4af22e58b69c0aa3a3221d70e9c/awx/main/fields.py#L615

```regex
^template(\.[a-zA-Z_]+[a-zA-Z0-9_]*)?$
```

Disclaimer: I _think_ that is right :)

File: awx/main/fields.py [hunk 3]
```diff
@@ -740,8 +739,24 @@ def validate(self, value, model_instance):
 
         class TowerNamespace:
             filename = None
-
         valid_namespace['tower'] = TowerNamespace()
+
+        # ensure either single file or multi-file syntax is used (but not both)
+        template_names = set(key for type_, injector in value.items()
```
Comment (ryanpetrello) [inline awx/main/fields.py]: This double-nested loop is a little hard to parse.  You're also needlessly looping over other injectors (like `env`) that don't actually _support_ the `template` key.  Can you just look at the `file` key instead e.g.,

```python
template_names = [x for x in value.get('file', {}).keys() if x.startswith('template')]
```

File: awx/main/models/credential.py [hunk 4]
```diff
@@ -553,17 +553,25 @@ class TowerNamespace:
...
+        for file_label, file_tmpl in file_tmpls.items():
             data = Template(file_tmpl).render(**namespace)
             _, path = tempfile.mkstemp(dir=private_data_dir)
             with open(path, 'w') as f:
                 f.write(data)
             os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)
-            namespace['tower'].filename = path
+
+            # determine if filename indicates single file or many
+            if file_label.find('.') == -1:
```
Comment (ryanpetrello) [inline awx/main/models/credential.py]: `if '.' not in file_label:`

or maybe even just:

`if file_label == 'template':`

File: awx/main/fields.py (hunk 1 above)
Comment (ryanpetrello) [reply awx/main/fields.py:765]: 👍 this makes sense to me, and uses the same validation approach as the old single `template` usage.

**Comment by ryanpetrello** [comment]
This is looking really good 💯.  The only thing I'd like to see is some functional tests that run through examples of task execution using multiple files.  This might be a good place to start:

https://github.com/ansible/tower/blob/fa09d68603e80e33fb28357c19ef9ec1f5d2a1a9/awx/main/tests/unit/test_tasks.py#L968

**Comment by jladdjr** [comment]
Awesome, appreciate all the feedback - will get some fixes in and leave an update when everything's been addressed. Thanks!

**Comment by wenottingham** [comment]
@jladdjr what is this waiting on?

File: awx/main/fields.py (hunk 2 above)
Comment (jladdjr) [reply awx/main/fields.py]: @ryanpetrello - is the `template.*` more of a convention instead of a requirement, then? It seemed like the `template.*` syntax was a requirement. (I definitely have it hard-coded elsewhere)

Comment (jladdjr) [reply awx/main/fields.py]: Since I'm setting the schema for the `file` field, specifically, it seemed worth enforcing the syntax here. It sounds like I might be missing something though.

Comment (ryanpetrello) [reply awx/main/fields.py]: I'm saying that it's possible for people to use jinja2 variables in their injectors that contain more complicated strings than those matched by `[a-zA-Z_]+`

What if you had an injectors like this?

```
"injectors": {
    "file": {
        "template.cert_1": "{{cert}}",
        "template.key_1": "{{key}}"
    },
    "env": {
        "MY_CERT": "{{tower.filename.cert_1}}",
        "MY_KEY": "{{tower.filename.key_1}}"
    }
}
```

The regex that I referenced at https://github.com/ansible/awx/blob/96904968d850c4af22e58b69c0aa3a3221d70e9c/awx/main/fields.py#L615 is more inclusive of valid variable names that include numbers, underscore, etc...

**Comment by jladdjr** [comment]
@wenottingham - nothing, just haven't had a chance to dust this off since break (the QE team is playing catch up). I think only a few minor changes are needed before this can be in.

**Comment by ryanpetrello** [comment]
What @jladdjr said - just a bit of polish and then this is probably ready to test.

File: awx/main/fields.py (hunk 3 above)
Comment (jladdjr) [reply awx/main/fields.py]: Will give that a try - now that I look at the double for loop again, I can't even make out what I was doing. Hard to read is right.

**Comment by ryanpetrello** [comment]
@jladdjr one last ask - if you'll add a test for the task executor:

https://github.com/ansible/awx/pull/696#issuecomment-347872629

...then I'm ready to merge this.

**Review by ryanpetrello — APPROVED** [review]
